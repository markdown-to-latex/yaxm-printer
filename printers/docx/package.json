{
    "name": "@md-to-latex/yaxm-printer-docx",
    "version": "0.0.0-beta.9",
    "description": "DOCX Printer for Yet Another eXtended Markdown",
    "main": "dist/index.js",
    "scripts": {
        "prebuild": "npm run convert-schema",
        "build": "tsc --build --pretty",
        "watch": "tsc --build --watch --pretty",
        "test": "jest test",
        "convert-schema": "node ./utils/convert-schema.js && prettier --write src/config/types.ts"
    },
    "repository": {
        "type": "git",
        "url": "https://github.com/markdown-to-latex/yaxm-printers"
    },
    "keywords": [
        "markdown",
        "latex",
        "converter",
        "transpiler",
        "report"
    ],
    "files": [
        "dist/**",
        "*.schema.json"
    ],
    "author": "Anatolii Titov <a@toliak.ru>",
    "license": "MIT",
    "devDependencies": {
        "@md-to-latex/converter": "^2.0.0-beta.13",
        "@types/jest": "^28.1.7",
        "@types/js-yaml": "^4.0.5",
        "@types/node": "^18.7.8",
        "jest": "^28.1.3",
        "json-schema-to-typescript": "^11.0.2",
        "prettier": "^2.7.1",
        "ts-jest": "^28.0.7",
        "ts-node": "^10.9.1",
        "typescript": "^4.7.4",
        "utils__": "file:../../utils"
    },
    "dependencies": {
        "convert-svg-to-png": "^0.6.4",
        "docx": "^7.8.2",
        "js-yaml": "^4.1.0",
        "json5": "^2.2.1",
        "jsonschema": "^1.4.1",
        "texsvg": "^2.0.1",
        "image-size": "^1.0.2"
    }
}
