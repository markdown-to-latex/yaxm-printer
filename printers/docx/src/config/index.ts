import { Validator, ValidatorResult } from 'jsonschema';
import * as path from 'path';
import * as JSON5 from 'json5';
import * as fs from 'fs';

export function validateConfig(data: unknown): ValidatorResult {
    const schemaPath = path.resolve(
        __dirname,
        '../../yaxm-printer.schema.json',
    );
    const schemaContent = fs.readFileSync(schemaPath, 'utf8');
    const schema = JSON5.parse(schemaContent);

    const validator = new Validator();
    return validator.validate(data, schema, {
        throwError: false,
    });
}

/*

export function readConfig(filepath: string): YAXMLatexPrinterConfig {
    const content = fs.readFileSync(filepath, 'utf8');
    const config = yaml.load(content) as MarkDownToLaTeXConverter;

    const validationResult = validateConfig(config);
    if (validationResult.errors.length > 0) {
        console.log(`Error happen while validating "${filepath}".`);
        throw new ConfigReaderError(
            JSON.stringify(validationResult.errors.map(e => e.toString())),
        );
    }

    return config;
}
*/

// --- api

export * from './types';
