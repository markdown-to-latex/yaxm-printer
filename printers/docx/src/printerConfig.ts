/**
 * Information about LaTeX margins
 */
export interface LatexPrinterConfigurationMarginInfo {
    /**
     * Figure: Inner Text Separator
     */
    imageInnerTextSep: string;

    /**
     * Figure: Below Caption Skip
     */
    imageBelowCaptionSkip: string;

    /**
     * Figure: Removed Below Caption Skip
     */
    imageRemovedBelowCaptionSkip: string;

    /**
     * Figure: Above Caption Skip
     */
    imageAboveCaptionSkip: string;

    /**
     * Figure: Inner Text Separator
     */
    codeInnerTextSep: string;

    /**
     * Figure: Below Caption Skip
     */
    codeBelowCaptionSkip: string;

    /**
     * Figure: Removed Below Caption Skip
     */
    codeRemovedBelowCaptionSkip: string;

    /**
     * Figure: Above Caption Skip
     */
    codeAboveCaptionSkip: string;

    /**
     * LongTable: Below Caption Skip
     */
    tableBelowCaptionSkip: string;

    /**
     * LongTable: Above Caption Skip
     */
    tableAboveCaptionSkip: string;

    /**
     * LongTable: Pre
     */
    tablePre: string;

    /**
     * LongTable: Post
     */
    tablePost: string;

    /**
     * LongTable: Removed Post Space
     */
    tableRemovedPost: string;

    /**
     * Code: Above Display Skip
     */
    mathAboveDisplaySkip: string;

    /**
     * Code: Below Display Skip
     */
    mathBelowDisplaySkip: string;

    /**
     * Code: Above Display Short Skip
     */
    mathAboveDisplayShortSkip: string;

    /**
     * Code: Below Display Short Skip
     */
    mathBelowDisplayShortSkip: string;
}

export type LatexInterpretation =
    | "default"
    | "monospace"
    | "bold"
    | "underline"
    | "italic"
    | "quotes";

export interface DocxPrinterConfiguration {
    useLinkAs: LatexInterpretation;
    useCodeSpanAs: LatexInterpretation;

    margin: LatexPrinterConfigurationMarginInfo;
}


export interface DocxPrinterConfigurationPartial
    extends Omit<
        Partial<DocxPrinterConfiguration>,
        "extendAutoEscapes" | "margin"
    > {
    margin?: Partial<LatexPrinterConfigurationMarginInfo>;
}

const defaultConfig: DocxPrinterConfiguration = {
    useCodeSpanAs: "quotes",
    useLinkAs: "default",
    margin: {
        imageInnerTextSep: "3em",
        imageBelowCaptionSkip: "-4ex",
        imageRemovedBelowCaptionSkip: "-1.6em",
        imageAboveCaptionSkip: "0.5em",
        codeInnerTextSep: "3em",
        codeBelowCaptionSkip: "-4ex",
        codeRemovedBelowCaptionSkip: "-1.6em",
        codeAboveCaptionSkip: "-0.5em",
        tableBelowCaptionSkip: "0em",
        tableAboveCaptionSkip: "0em",
        tablePre: "2em",
        tablePost: "2em",
        tableRemovedPost: "0em",
        mathAboveDisplaySkip: "-0.9em",
        mathBelowDisplaySkip: "0pt",
        mathAboveDisplayShortSkip: "0pt",
        mathBelowDisplayShortSkip: "0pt"
    }
};

export function buildConfig(
    config?: DocxPrinterConfigurationPartial
): DocxPrinterConfiguration {
    return {
        useCodeSpanAs: config?.useCodeSpanAs ?? defaultConfig.useCodeSpanAs,
        useLinkAs: config?.useLinkAs ?? defaultConfig.useLinkAs,
        margin: {
            imageInnerTextSep:
                config?.margin?.imageInnerTextSep ??
                defaultConfig.margin.imageInnerTextSep,
            imageBelowCaptionSkip:
                config?.margin?.imageBelowCaptionSkip ??
                defaultConfig.margin.imageBelowCaptionSkip,
            imageRemovedBelowCaptionSkip:
                config?.margin?.imageRemovedBelowCaptionSkip ??
                defaultConfig.margin.imageRemovedBelowCaptionSkip,
            imageAboveCaptionSkip:
                config?.margin?.imageAboveCaptionSkip ??
                defaultConfig.margin.imageAboveCaptionSkip,
            codeInnerTextSep:
                config?.margin?.codeInnerTextSep ??
                defaultConfig.margin.codeInnerTextSep,
            codeBelowCaptionSkip:
                config?.margin?.codeBelowCaptionSkip ??
                defaultConfig.margin.codeBelowCaptionSkip,
            codeRemovedBelowCaptionSkip:
                config?.margin?.codeRemovedBelowCaptionSkip ??
                defaultConfig.margin.codeRemovedBelowCaptionSkip,
            codeAboveCaptionSkip:
                config?.margin?.codeAboveCaptionSkip ??
                defaultConfig.margin.codeAboveCaptionSkip,
            tableBelowCaptionSkip:
                config?.margin?.tableBelowCaptionSkip ??
                defaultConfig.margin.tableBelowCaptionSkip,
            tableAboveCaptionSkip:
                config?.margin?.tableAboveCaptionSkip ??
                defaultConfig.margin.tableAboveCaptionSkip,
            tablePre: config?.margin?.tablePre ?? defaultConfig.margin.tablePre,
            tablePost:
                config?.margin?.tablePost ?? defaultConfig.margin.tablePost,
            tableRemovedPost:
                config?.margin?.tableRemovedPost ??
                defaultConfig.margin.tableRemovedPost,
            mathAboveDisplaySkip:
                config?.margin?.mathAboveDisplaySkip ??
                defaultConfig.margin.mathAboveDisplaySkip,
            mathBelowDisplaySkip:
                config?.margin?.mathBelowDisplaySkip ??
                defaultConfig.margin.mathBelowDisplaySkip,
            mathAboveDisplayShortSkip:
                config?.margin?.mathAboveDisplayShortSkip ??
                defaultConfig.margin.mathAboveDisplayShortSkip,
            mathBelowDisplayShortSkip:
                config?.margin?.mathBelowDisplayShortSkip ??
                defaultConfig.margin.mathBelowDisplayShortSkip
        }
    };
}
