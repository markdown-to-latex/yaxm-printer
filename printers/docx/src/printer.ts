import {
    ListItemNode,
    ListNode,
    NodeText,
} from '@md-to-latex/converter/dist/ast/node';
import { DiagnoseList } from '@md-to-latex/converter/dist/diagnostic';
import * as docx from 'docx';
import {
    AlignmentType,
    Paragraph,
    TextRun,
    VerticalAlign,
    WidthType,
} from 'docx';
import { default as texsvg } from 'texsvg';
import { YAMLException } from 'js-yaml';

// @ts-ignore
import * as svg2png from 'convert-svg-to-png';
import { FormulaProcessedNode } from '@md-to-latex/converter/dist/macro/node';
import { DocxPrinter } from './index';

export interface PrinterFunctionResult {
    result: docx.XmlComponent[];
    diagnostic: DiagnoseList;
}

function getOrCreateWordListRef(
    printer: DocxPrinter,
    node: ListNode,
    isOrdered: boolean,
): string {
    // wow https://stackoverflow.com/a/12502559/14142236
    const randomRef = () => Math.random().toString(36).slice(2);

    const ref = printer.wordListRefStore.find(
        entry => entry.node === node,
    )?.ref;
    if (ref) {
        return ref;
    }

    const newRef = randomRef();
    printer.wordListRefStore.push({ node, ref: newRef, isOrdered });
    return newRef;
}

export interface GetWordListItemData {
    xml: docx.XmlComponent[];
    depth: number;
    index: number;
    isOrdered: boolean;
}

export async function getWordListItem(
    printer: DocxPrinter,
    data: GetWordListItemData,
    node: ListItemNode,
    superparentList: ListNode,
): Promise<PrinterFunctionResult> {
    const ref = getOrCreateWordListRef(
        printer,
        superparentList,
        data.isOrdered,
    );
    return {
        result: [
            new Paragraph({
                children: data.xml,
                numbering: {
                    instance: 0,
                    reference: ref,
                    level: data.depth - 1,
                },
            }),
        ],
        diagnostic: [],
    };
}

const RE_SVG_SIZE_W = /<svg [^>]+ width="([0-9.]+)ex" [^>]+>/g;
const RE_SVG_SIZE_H = /<svg [^>]+ height="([0-9.]+)ex" [^>]+>/g;

export function getSvgSizeEx(svgRaw: string): [number, number] {
    const matchW: RegExpMatchArray | undefined = svgRaw
        .matchAll(RE_SVG_SIZE_W)
        .next().value;
    if (!matchW) {
        // TODO: diagnostic
        throw new YAMLException('Unable to get width');
    }
    const matchH: RegExpMatchArray | undefined = svgRaw
        .matchAll(RE_SVG_SIZE_H)
        .next().value;
    if (!matchH) {
        // TODO: diagnostic
        throw new YAMLException('Unable to get height');
    }

    return [+matchW[1], +matchH[1]];
}

const RE_SVG_ERROR = /data-mjx-error="([^"]+)"/g;

export async function printKeyNode(node: { index: number }) {
    return {
        result: [
            new docx.TextRun({
                text: `${node.index + 1}`,
            }),
        ],
        diagnostic: [],
    };
}

export async function printLazyNumberNode(fun: () => number) {
    return {
        result: [
            new docx.TextRun({
                text: `${fun()}`,
            }),
        ],
        diagnostic: [],
    };
}

// TODO: with diagnostic
export async function printFormulaProcessedNode(
    node: FormulaProcessedNode,
): Promise<docx.XmlComponent> {
    return new docx.Table({
        borders: {
            right: { style: docx.BorderStyle.NONE },
            left: { style: docx.BorderStyle.NONE },
            top: { style: docx.BorderStyle.NONE },
            bottom: { style: docx.BorderStyle.NONE },
            insideVertical: { style: docx.BorderStyle.NONE },
            insideHorizontal: { style: docx.BorderStyle.NONE },
        },
        rows: [
            new docx.TableRow({
                children: [
                    new docx.TableCell({
                        children: [
                            new docx.Paragraph({
                                children: [
                                    await formulaNodeToPicture(node.text),
                                ],
                                alignment: AlignmentType.CENTER,
                            }),
                        ],
                        verticalAlign: VerticalAlign.CENTER,
                    }),
                    new docx.TableCell({
                        children: [
                            new docx.Paragraph({
                                text: `(${node.index + 1})`,
                                alignment: AlignmentType.RIGHT,
                            }),
                        ],
                        verticalAlign: VerticalAlign.CENTER,
                        width: {
                            type: WidthType.DXA,
                            size: 510, // is 0.35 inches == 0.9 cm
                        },
                    }),
                ],
            }),
        ],
        width: {
            type: WidthType.PERCENTAGE,
            size: 100,
        },
    });
}

// TODO: with diagnostic
export async function formulaNodeToPicture(
    node: NodeText,
): Promise<docx.ImageRun> {
    let text = `\\begin{align*}${node.text}\\end{align*}`;

    let svgRaw;
    try {
        svgRaw = await texsvg(text);
    } catch (e: any) {
        console.error(`Error: ${e}`);
        const captured = (() => {
            const value = (e.source ?? '').matchAll(RE_SVG_ERROR).next().value;
            if (!value) {
                return undefined;
            }
            return value[1] as string;
        })();
        const message = captured ? captured : 'Unknown';

        // TODO: diagnostic
        throw new YAMLException(`Unable to convert tex to svg: ${message}`);
    }

    const [w, h] = getSvgSizeEx(svgRaw);
    const k = 8;

    let outputBuffer: Buffer = await svg2png.convert(svgRaw, {
        scale: 100,
    });

    console.log(`W: ${w} H: ${h}`);

    return new docx.ImageRun({
        data: outputBuffer,
        transformation: {
            width: Math.ceil(w * k),
            height: Math.ceil(h * k),
        },
    });
}

export interface WordTableInfo {
    tableIndex: string;
    tableTitle: docx.XmlComponent[];
    header: docx.TableRow;
    content: docx.TableRow[];
    colAmount: number;
}

export function getWordTable(info: WordTableInfo): PrinterFunctionResult {
    return {
        result: [
            new docx.Paragraph({
                children: [
                    new docx.TextRun({
                        text: `Таблица ${info.tableIndex} – `,
                    }),
                    ...info.tableTitle,
                ],
            }),
            new docx.Table({
                width: {
                    type: WidthType.PERCENTAGE,
                    size: 100,
                },
                rows: [info.header, ...info.content],
            }),
        ],
        diagnostic: [],
    };
}

// ---------

export interface IRunOptionsExt extends docx.IRunOptions {
    /**
     * In pt/2
     */
    position?: number;
    spacing?: number;
}

export function createTextRunExt(options: IRunOptionsExt | string) {
    // export class TextRunExt extends docx.TextRun
    // doesn't work due to runtime constructor error

    const textRun = new docx.TextRun(options);
    const properties = (textRun as any).properties as docx.RunProperties;
    if (typeof options !== 'string') {
        if (options.position) {
            properties.push(
                new docx.NumberValueElement('w:position', options.position),
            );
        }
        if (options.spacing) {
            properties.push(
                new docx.NumberValueElement('w:spacing', options.spacing),
            );
        }
    }

    return textRun;
}
