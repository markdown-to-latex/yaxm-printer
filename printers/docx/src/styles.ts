import * as docx from 'docx';
import { AlignmentType } from 'docx';

function fontSizeToDocxFontSize(size: number): number {
    return size * 2;
}

export function getDocumentGlobalStyles(): docx.IStylesOptions {
    const defaultSpacing = {
        after: 0,
        before: 0,
        // I don't know, it's 1.5 spacing
        line: 360,
    };
    const defaultIndent = {
        firstLine: '1.25cm',
    };
    const defaultFont = {
        font: 'Tinos',
        size: fontSizeToDocxFontSize(14),
    };
    const defaultPictureSpacingMm = 6;
    const defaultTableSpacingMm = defaultPictureSpacingMm * 1.5;

    return {
        default: {
            document: {
                paragraph: {
                    alignment: AlignmentType.JUSTIFIED,
                    indent: { ...defaultIndent },
                    spacing: { ...defaultSpacing },
                },
                run: { ...defaultFont },
            },
            heading1: {
                paragraph: {
                    numbering: {
                        reference: HEADING_REF_NAME,
                        level: 0,
                    },
                    alignment: AlignmentType.CENTER,
                    indent: { firstLine: 0 },
                    spacing: { ...defaultSpacing },
                    keepNext: true,
                },
                run: { ...defaultFont, bold: true, allCaps: true },
            },
            heading2: {
                paragraph: {
                    numbering: {
                        reference: HEADING_REF_NAME,
                        level: 1,
                    },
                    alignment: AlignmentType.JUSTIFIED,
                    indent: { ...defaultIndent },
                    spacing: { ...defaultSpacing },
                    keepNext: true,
                },
                run: { ...defaultFont, bold: true },
            },
            heading3: {
                paragraph: {
                    numbering: {
                        reference: HEADING_REF_NAME,
                        level: 2,
                    },
                    alignment: AlignmentType.JUSTIFIED,
                    indent: { ...defaultIndent },
                    spacing: { ...defaultSpacing },
                    keepNext: true,
                },
                run: { ...defaultFont, bold: true },
            },
        },
        paragraphStyles: [
            // Default
            {
                id: 'tocheading',
                name: 'TOC Heading',
                paragraph: {
                    alignment: AlignmentType.CENTER,
                    indent: { firstLine: 0 },
                    spacing: { ...defaultSpacing },
                    keepNext: true,
                },
                run: { ...defaultFont, bold: true, allCaps: true },
            },
            {
                id: 'toc 1',
                name: 'TOC 1',
                paragraph: {
                    indent: { firstLine: 0, start: 0 },
                    spacing: { ...defaultSpacing },
                },
                run: { ...defaultFont, allCaps: true },
            },
            {
                id: 'toc 2',
                name: 'TOC 2',
                paragraph: {
                    indent: {
                        firstLine: docx.convertMillimetersToTwip(12.5),
                        start: 0,
                    },
                    spacing: { ...defaultSpacing },
                },
                run: { ...defaultFont },
            },
            {
                id: 'toc 3',
                name: 'TOC 3',
                paragraph: {
                    indent: {
                        firstLine: docx.convertMillimetersToTwip(25.0),
                        start: 0,
                    },
                    spacing: { ...defaultSpacing },
                },
                run: { ...defaultFont },
            },

            // Custom
            {
                id: 'code',
                name: 'code',
                paragraph: {
                    spacing: {
                        before: docx.convertMillimetersToTwip(
                            defaultTableSpacingMm,
                        ),
                    },
                    keepLines: true,
                    keepNext: true,
                    indent: {
                        firstLine: 0,
                    },
                    alignment: AlignmentType.LEFT,
                },
                run: {
                    size: fontSizeToDocxFontSize(12),
                    font: 'Fira Code',
                },
            },
            {
                id: 'table-caption',
                name: 'table-caption',
                run: {},
                paragraph: {
                    spacing: {
                        before: docx.convertMillimetersToTwip(
                            defaultPictureSpacingMm,
                        ),
                    },
                    indent: {
                        firstLine: 0,
                    },
                },
            },
            {
                id: 'table',
                name: 'table',
                paragraph: {
                    spacing: {
                        after: docx.convertMillimetersToTwip(
                            defaultTableSpacingMm,
                        ),
                    },
                    indent: {
                        firstLine: 0,
                    },
                },
            },
            {
                id: 'table-cell',
                name: 'table-cell',
                paragraph: {
                    alignment: AlignmentType.CENTER,
                    indent: {
                        firstLine: 0,
                    },
                },
            },
            {
                id: 'picture-caption',
                name: 'picture-caption',
                run: {},
                paragraph: {
                    alignment: AlignmentType.CENTER,
                    spacing: {
                        after: docx.convertMillimetersToTwip(
                            defaultPictureSpacingMm,
                        ),
                    },
                    indent: {
                        firstLine: 0,
                    },
                },
            },
            {
                id: 'picture',
                name: 'picture',
                paragraph: {
                    keepNext: true,
                    alignment: AlignmentType.CENTER,
                    spacing: {
                        before: docx.convertMillimetersToTwip(
                            defaultPictureSpacingMm,
                        ),
                    },
                    indent: {
                        firstLine: 0,
                    },
                },
            },
            {
                id: 'formula-picture',
                name: 'formula-picture',
                paragraph: {
                    alignment: AlignmentType.CENTER,
                    spacing: {
                        before: docx.convertMillimetersToTwip(
                            defaultPictureSpacingMm,
                        ),
                        after: docx.convertMillimetersToTwip(
                            defaultPictureSpacingMm,
                        ),
                    },
                    indent: {
                        firstLine: 0,
                    },
                },
            },
            {
                id: 'formula-table-cell',
                name: 'formula-table-cell',
                paragraph: {
                    alignment: AlignmentType.CENTER,
                    indent: {
                        firstLine: 0,
                    },
                },
            },
            {
                id: 'formula-table-cell-number',
                name: 'formula-table-cell-number',
                paragraph: {
                    alignment: AlignmentType.CENTER,
                    indent: {
                        firstLine: 0,
                    },
                },
            },
        ],
    };
}

interface NumberingLevelSpacing {
    tabStop: number;
    firstLine: number;
}

const numberingLevelsSpacings = (level: number) =>
    ({
        tabStop: docx.convertMillimetersToTwip(level * 15 + 10),
        firstLine: docx.convertMillimetersToTwip(level * 15),
    } as NumberingLevelSpacing);

export function getOrderedNumberingLevels(): docx.ILevelsOptions[] {
    return [
        {
            level: 0,
            format: docx.LevelFormat.RUSSIAN_LOWER,
            text: '%1)',
            alignment: docx.AlignmentType.START,
            style: {
                paragraph: {
                    leftTabStop: numberingLevelsSpacings(1).tabStop,
                    indent: {
                        firstLine: numberingLevelsSpacings(1).firstLine,
                    },
                },
            },
        },
        {
            level: 1,
            format: docx.LevelFormat.DECIMAL,
            text: '%2)',
            alignment: docx.AlignmentType.START,
            style: {
                paragraph: {
                    leftTabStop: numberingLevelsSpacings(2).tabStop,
                    indent: {
                        firstLine: numberingLevelsSpacings(2).firstLine,
                    },
                },
            },
        },
        {
            level: 2,
            format: docx.LevelFormat.UPPER_ROMAN,
            text: '%3)',
            alignment: docx.AlignmentType.START,
            style: {
                paragraph: {
                    leftTabStop: numberingLevelsSpacings(3).tabStop,
                    indent: {
                        firstLine: numberingLevelsSpacings(3).firstLine,
                    },
                },
            },
        },
    ];
}

export function getUnorderedNumberingLevels(): docx.ILevelsOptions[] {
    return [
        {
            level: 0,
            format: docx.LevelFormat.NONE,
            text: '-',
            alignment: docx.AlignmentType.START,
            style: {
                paragraph: {
                    leftTabStop: numberingLevelsSpacings(1).tabStop,
                    indent: {
                        firstLine: numberingLevelsSpacings(1).firstLine,
                    },
                },
            },
        },
        {
            level: 1,
            format: docx.LevelFormat.RUSSIAN_LOWER,
            text: '%2)',
            alignment: docx.AlignmentType.START,
            style: {
                paragraph: {
                    leftTabStop: numberingLevelsSpacings(2).tabStop,
                    indent: {
                        firstLine: numberingLevelsSpacings(2).firstLine,
                    },
                },
            },
        },
        {
            level: 2,
            format: docx.LevelFormat.DECIMAL,
            text: '%3)',
            alignment: docx.AlignmentType.START,
            style: {
                paragraph: {
                    leftTabStop: numberingLevelsSpacings(3).tabStop,
                    indent: {
                        firstLine: numberingLevelsSpacings(3).firstLine,
                    },
                },
            },
        },
    ];
}

export type INumberingOptionsConfig = docx.INumberingOptions['config'][0];

const HEADING_REF_NAME = 'heading-ref';

export function getNumberingHeading(): INumberingOptionsConfig {
    return {
        reference: HEADING_REF_NAME,
        levels: [
            {
                level: 0,
                format: docx.LevelFormat.NONE,
                text: '',
                alignment: docx.AlignmentType.START,
                style: {
                    paragraph: {
                        leftTabStop: docx.convertMillimetersToTwip(10),
                    },
                },
            },
            {
                level: 1,
                format: docx.LevelFormat.DECIMAL,
                text: '%2',
                alignment: docx.AlignmentType.START,
                style: {
                    paragraph: {
                        leftTabStop: docx.convertMillimetersToTwip(20 + 15),
                    },
                },
            },
            {
                level: 2,
                format: docx.LevelFormat.DECIMAL,
                text: '%2.%3',
                alignment: docx.AlignmentType.START,
                style: {
                    paragraph: {
                        leftTabStop: docx.convertMillimetersToTwip(20 + 15),
                    },
                },
            },
        ],
    };
}
