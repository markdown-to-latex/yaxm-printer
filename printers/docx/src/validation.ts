import * as docx from "docx";
import { XmlComponent } from "docx";
import {
    DiagnoseErrorType,
    DiagnoseInfo,
    DiagnoseList,
    DiagnoseSeverity
} from "@md-to-latex/converter/dist/diagnostic";

function docxRootDiagnose(severity: DiagnoseSeverity, message: string): DiagnoseInfo {
    return {
        pos: { start: { line: 0, column: 0, absolute: 0 }, end: { line: 0, column: 0, absolute: 0 } },
        errorType: DiagnoseErrorType.OtherError,
        filePath: ".",
        message,
        severity
    };
}

function getXmlComponentRoot(docxNode: Readonly<XmlComponent>): any[] {
    interface XmlComponentWithRoot {
        root?: any[];
    }

    let casted = docxNode as unknown as XmlComponentWithRoot;

    if (!casted.root) {
        return [];
    }
    return casted.root;
}

export function validateDocxRootNode(docxNode: Readonly<XmlComponent>): DiagnoseList {
    let diagnostic: DiagnoseList = [];

    if (!(docxNode instanceof docx.Paragraph)) {
        // TODO: provide more details: which node, etc..
        diagnostic.push(docxRootDiagnose(DiagnoseSeverity.Error, "The docx root node is not a paragraph"));
    }

    let queue: Readonly<XmlComponent>[] = [docxNode];
    while (queue.length !== 0) {
        let component = queue.pop()!;

        // TODO: idk why the root is children wth
        let children = getXmlComponentRoot(component);
        for (let child of children) {
            if (child instanceof docx.Paragraph) {
                diagnostic.push(
                    docxRootDiagnose(
                        DiagnoseSeverity.Error,
                        "Docx Paragraph in paragraph detected"
                    )
                );
            }
        }

        queue.push(...children);
    }


    return diagnostic;
}