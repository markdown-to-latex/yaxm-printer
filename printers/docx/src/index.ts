import {
    DocxPrinterVisitor,
    DocxPrinterVisitorList,
    ProcessingVisitors,
    processingVisitors,
} from './visitors';
import { buildConfig, DocxPrinterConfiguration } from './printerConfig';
import { NodeProcessed } from '@md-to-latex/converter/dist/macro/node';
import { DiagnoseList } from '@md-to-latex/converter/dist/diagnostic';
import { ListNode, Node, RawNode } from '@md-to-latex/converter/dist/ast/node';
import * as docx from 'docx';
import { XmlComponent } from 'docx';
import { validateDocxRootNode } from './validation';
import {
    getDocumentGlobalStyles,
    getNumberingHeading,
    getOrderedNumberingLevels,
    getUnorderedNumberingLevels,
    INumberingOptionsConfig,
} from './styles';

const processNode: DocxPrinterVisitor<NodeProcessed | RawNode | Node> =
    function (printer, node) {
        return processingVisitors[node.type](
            printer,
            /* TODO: resolve that */ node as any,
        );
    };

const processNodeList: DocxPrinterVisitorList<NodeProcessed | RawNode | Node> =
    async function (printer, nodes) {
        const processedNodes = await Promise.all(
            nodes.map(async node => await printer.processNode(printer, node)),
        );

        const diagnostic: DiagnoseList = processedNodes.flatMap(
            nodes => nodes.diagnostic,
        );
        const result = processedNodes.flatMap(nodes => nodes.result);

        return {
            result,
            diagnostic,
        };
    };

export interface WordRefData {
    node: ListNode;
    ref: string;
    isOrdered: boolean;
}

export interface DocxPrinter {
    processNode: DocxPrinterVisitor<NodeProcessed | RawNode | Node>;
    processNodeList: DocxPrinterVisitorList<NodeProcessed | RawNode | Node>;

    processingVisitors: ProcessingVisitors;
    config: DocxPrinterConfiguration;

    wordListRefStore: WordRefData[];
}

export interface PrinterVisitorResult {
    result: docx.XmlComponent[];
    diagnostic: DiagnoseList;
}

export function createPrinterDocx(
    config: DocxPrinterConfiguration,
): DocxPrinter {
    return {
        processingVisitors,
        config: config,

        processNode,
        processNodeList,

        wordListRefStore: [],
    };
}

export async function printerResultToBuffer(
    printer: DocxPrinter,
    result: Readonly<XmlComponent[]>,
): Promise<Buffer> {
    console.log(result);

    const doc = new docx.Document({
        styles: getDocumentGlobalStyles(),
        numbering: {
            config: [
                ...printer.wordListRefStore.map<INumberingOptionsConfig>(n =>
                    n.isOrdered
                        ? {
                              reference: n.ref,
                              levels: getOrderedNumberingLevels(),
                          }
                        : {
                              reference: n.ref,
                              levels: getUnorderedNumberingLevels(),
                          },
                ),
                getNumberingHeading(),
            ],
        },
        features: {
            updateFields: true,
        },
        sections: [
            {
                properties: {
                    page: {
                        size: {
                            orientation: docx.PageOrientation.PORTRAIT,
                            height: docx.convertMillimetersToTwip(297),
                            width: docx.convertMillimetersToTwip(210),
                        },
                        margin: {
                            top: docx.convertMillimetersToTwip(20),
                            right: docx.convertMillimetersToTwip(10),
                            bottom: docx.convertMillimetersToTwip(20),
                            left: docx.convertMillimetersToTwip(30),
                        },
                    },
                },
                footers: {
                    default: new docx.Footer({
                        children: [
                            new docx.Paragraph({
                                indent: {
                                    firstLine: 0,
                                },
                                alignment: docx.AlignmentType.CENTER,
                                children: [
                                    new docx.TextRun({
                                        children: [docx.PageNumber.CURRENT],
                                    }),
                                ],
                            }),
                        ],
                    }),
                },
                children: [
                    ...result.map(n => {
                        if (
                            !(
                                n instanceof docx.Paragraph ||
                                n instanceof docx.Table
                            )
                        ) {
                            return new docx.Paragraph({
                                children: [n],
                            });
                        }
                        return n;
                    }),
                ],
            },
        ],
    });
    return await docx.Packer.toBuffer(doc);
}

export { buildConfig, validateDocxRootNode };
