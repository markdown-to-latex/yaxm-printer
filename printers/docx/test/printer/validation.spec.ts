import * as docx from "docx";
import { validateDocxRootNode } from "../../src/validation";

describe("validation test", function() {
    test("paragraph must be the root node", function() {
        let diagnostic = validateDocxRootNode(new docx.TextRun({}));
        expect(diagnostic).toMatchSnapshot();
    });

    test("paragraph must not contain paragraphs", function() {
        let node = new docx.Paragraph({
            children: [
                new docx.TextRun({
                    children: [new docx.Paragraph({})]
                })
            ]
        });
        let diagnostic = validateDocxRootNode(node);
        expect(diagnostic).toMatchSnapshot();
    });
});