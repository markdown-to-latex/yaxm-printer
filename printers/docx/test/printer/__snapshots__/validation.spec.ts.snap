// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation test paragraph must be the root node 1`] = `
Array [
  Object {
    "errorType": 7,
    "filePath": ".",
    "message": "The docx root node is not a paragraph",
    "pos": Object {
      "end": Object {
        "absolute": 0,
        "column": 0,
        "line": 0,
      },
      "start": Object {
        "absolute": 0,
        "column": 0,
        "line": 0,
      },
    },
    "severity": "ERROR",
  },
]
`;

exports[`validation test paragraph must not contain paragraphs 1`] = `
Array [
  Object {
    "errorType": 7,
    "filePath": ".",
    "message": "Docx Paragraph in paragraph detected",
    "pos": Object {
      "end": Object {
        "absolute": 0,
        "column": 0,
        "line": 0,
      },
      "start": Object {
        "absolute": 0,
        "column": 0,
        "line": 0,
      },
    },
    "severity": "ERROR",
  },
]
`;
